### PR TITLE
Create initial provider

### PR DESCRIPTION
**Pull request by tptee**
Creates the initial `serverless_deployment` resource which wraps the Serverless CLI. Only runs `sls deploy` when the contents of `serverless.yml` and the package ZIP archive contents change. Assumes that users run packaging out-of-band.

Hoping to split docs and niceties into another PR while we dogfood the provider.

Example:

```hcl
resource "serverless_deployment" "example" {
  config_path         = abspath("./")
  serverless_bin_path = abspath("./node_modules/.bin/serverless")
  # Relative to config path
  package_dir         = ".terraform-serverless"
  stage               = "sandbox"
}
```

To test, you need access to the Formidable account set up with `aws-vault`. Run `aws-vault exec formidable` then `make testacc`.

@dmmulroy I'd love your input here since you have some experience with Go!

### CODE REVIEW

File: scripts/changelog-links.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+if [[ `uname` == "Darwin" ]]; then
+  echo "Using BSD sed"
+  SED="sed -i.bak -E -e"
+else
+  echo "Using GNU sed"
+  SED="sed -i.bak -r -e"
+fi
+
+PROVIDER_URL="https:\/\/github.com\/terraform-providers\/terraform-provider-fastly\/issues"
```
Comment (ryan-roemer) [inline scripts/changelog-links.sh]: Need to refactor `terraform-provider-fastly` and the repo string to this project.

File: serverless.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  runtime: nodejs10.x
+
+package:
+  include:
+    # Root level
+    - "!*"
+    - "package.json"
+    # Source directories
+    - "!aws/**"
+    - "!terraform"
```
Comment (ryan-roemer) [inline serverless.yml]: GLOBAL: I updated the issue/docs in jetpack, but, but you don't need:

```yaml
- "!terraform"
- "!terraform/**"
```

You only need:

```yaml
- "!terraform/**"
```

File: serverless/resource_deployment_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		}
+	}
+	return err
+}
+
+func testAccServerlessDeploymentConfig() string {
+	return `
+resource "serverless_deployment" "example" {
+  config_path         = abspath("../serverless.yml")
+  serverless_bin_path = abspath("../node_modules/.bin/serverless")
```
Comment (ryan-roemer) [inline serverless/resource_deployment_test.go]: Is this going to be the same on Windows and still executable? I usually see `{NAME}.(cmd|ps1)`

For example, in jetpack testing, we use this JavaScript:

```js
const IS_WIN = process.platform === "win32";
const SLS_CMD = `node_modules/.bin/serverless${IS_WIN ? ".cmd" : ""}`;
```

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,7 @@
+resource "serverless_deployment" "example" {
+  config_path         = abspath("serverless.yml")
+  # Relative to config path
+  package_dir         = ".terraform-serverless"
+  stage               = "sandbox"
+  serverless_bin_path = abspath("node_modules/.bin/serverless")
```
Comment (ryan-roemer) [inline main.tf]: Is this going to be the same on Windows and still executable? I usually see `{NAME}.(cmd|ps1)`

For example, in jetpack testing, we use this JavaScript:

```js
const IS_WIN = process.platform === "win32";
const SLS_CMD = `node_modules/.bin/serverless${IS_WIN ? ".cmd" : ""}`;
```

File: index.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+'use strict';
```
Comment (ryan-roemer) [inline index.js]: Maybe stick this in an `examples` directory to keep everything _not_ the provider easily separated?

File: serverless/resource_deployment.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+func resourceDeployment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDeploymentCreate,
+		Read:   resourceDeploymentRead,
+		Update: resourceDeploymentUpdate,
+		Delete: resourceDeploymentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"config_path": &schema.Schema{
```
Comment (ryan-roemer) [inline serverless/resource_deployment.go]: Serverless does this as automatic config inference: https://github.com/serverless/serverless/blob/1cad7b88fb92123e94ab2f5eae2e4f06ef7b3001/lib/utils/getServerlessConfigFile.js#L17-L39

In particular choosing config files based on this implied order:

```js
  const yamlPath = path.join(servicePath, 'serverless.yaml');
  const ymlPath = path.join(servicePath, 'serverless.yml');
  const jsonPath = path.join(servicePath, 'serverless.json');
  const jsPath = path.join(servicePath, 'serverless.js');
```

We could potentially offer a default option if not set:
- (TOTALLY OPTIONAL) Take `servicePath` arguments like SLS does.
- Do the same default config file order, which we should do for correctness if allowing a default.

... or require absolute config specification like I think we have now which is fine. But maybe document the above scenario just a bit?

Comment (tptee) [reply serverless/resource_deployment.go]: This actually exposes another problem–we parse the YAML and grab the service name from there, but not all Serverless configs are YAML! Will have a think on this one.

Comment (tptee) [reply serverless/resource_deployment.go]: Aha! I can use ` sls print --path service --format text`

**Comment by tptee** [comment]
Ready for another look!

File: index.js (hunk 5 above)
Comment (ryan-roemer) [reply index.js]: I think this still needs to get moved to examples?

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,7 @@
+resource "serverless_deployment" "example" {
```
Comment (ryan-roemer) [inline main.tf]: Move to example if we can?

File: index.js (hunk 5 above)
Comment (tptee) [reply index.js]: Done!

**Comment by tptee** [comment]
I believe the `push` Travis job will stop failing after merge, although I find the pr/push thing deeply confusing!

File: main.tf (hunk 7 above)
Comment (ryan-roemer) [reply main.tf]: Should/can we rename it to `example-test.tf` or something to clearly distinguish it?

Comment (tptee) [reply main.tf]: Done!

File: yarn.lock [hunk 8]
```diff
@@ -0,0 +1,3112 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (ryan-roemer) [inline yarn.lock]: I think this file needs to be removed (it's in `example` now).

File: serverless/resource_deployment.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+func getServiceName(configJson []byte) (string, error) {
+	config := serverlessConfig{}
+	err := json.Unmarshal([]byte(configJson), &config)
+
+	return config.Service, err
+}
+
+func hashServerlessDir(configDir string, packagePath string, configJson []byte) (string, error) {
```
Comment (ryan-roemer) [inline serverless/resource_deployment.go]: Can we get some comments as to what we're hashing, why, and what scenarios we're trying to cache / skip over?

File: serverless/resource_deployment.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	absolutePackagePath := filepath.Join(configDir, packagePath)
+	zipPath := filepath.Join(absolutePackagePath, "sls-provider.zip")
```
Comment (ryan-roemer) [inline serverless/resource_deployment.go]: This looks hard-coded. Isn't this supposed to be based on `getServiceName`?

Comment (ryan-roemer) [reply serverless/resource_deployment.go]: Also, what happens if `package.individually = true` is set and we get lots of zips?

File: serverless/resource_deployment.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+		Delete: resourceDeploymentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"config_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"serverless_bin_dir": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (ryan-roemer) [inline serverless/resource_deployment.go]: Do we want to make this not required and instead default to CWD?

File: serverless/resource_deployment.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+			},
+			"package_dir": &schema.Schema{
```
Comment (ryan-roemer) [inline serverless/resource_deployment.go]: What is `package_dir`? Can we get a comment here?

File: serverless/resource_deployment.go (hunk 11 above)
Comment (tptee) [reply serverless/resource_deployment.go]: My gut is that would be more error-prone, as I haven't seen a project share Terraform modules and `serverless.yml`/`node_modules` in the same CWD before.

Comment (tptee) [reply serverless/resource_deployment.go]: Oh you know what, might be nice to default it to the `node_modules` in the `config_dir`!

**Comment by ryan-roemer** [comment]
Looking great so far! In addition to my current review pass comments, can we maybe get a `CONTRIBUTING.md` document that has sections for:

- `Development`: Localdev installation, running the tests, running lint, doing a deploy from `example` (with installation, etc. instructions)
- `Release`: Since git tags should cause a full release, maybe just giving guidance for that via straight git commands since we don't have a root package.json for `scripts`? (Or something of your choosing).

Here's https://github.com/FormidableLabs/terraform-aws-serverless/blob/master/CONTRIBUTING.md for reference which has some of what you may want if that's useful!

File: serverless/resource_deployment.go (hunk 12 above)
Comment (tptee) [reply serverless/resource_deployment.go]: Added comment!

File: serverless/resource_deployment.go (hunk 11 above)
Comment (tptee) [reply serverless/resource_deployment.go]: Made the above the default!

File: serverless/resource_deployment.go (hunk 10 above)
Comment (tptee) [reply serverless/resource_deployment.go]: Whew glad you caught that hardcoding 😅 for the `package.individually`, I'd like to ticket it our for a follow-up PR. I imagine we can hash all of the zip files in `package_dir` to solve it!

**Comment by tptee** [comment]
Should be ready for another pass!

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+## Before submitting a PR...
+
+- run the acceptance tests! See the instructions above.
+- run `make fmt` to format your code! CI will catch this, but it'll save you a build!
+
+## Releases
+
+- Push a version tag (e.g. v1.2.3) to master after merging the PRs you want to release.
```
Comment (ryan-roemer) [inline CONTRIBUTING.md]: Let's do an example like `git tag -a "v1.2.3" -m "1.2.3"` in there just for ease of use?

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+Terraform infers log levels from string prefixes. To log values from the provider, you can use:
+
+```go
+log.Println("[WARN]", value)
+```
+
+You can then use `TF_LOG=warn` before `terraform apply` or `make testacc` to see your logs.
+
+## Testing
```
Comment (ryan-roemer) [inline CONTRIBUTING.md]: lint too?

File: serverless/resource_deployment.go (hunk 10 above)
Comment (tptee) [reply serverless/resource_deployment.go]: Ticket for `package.individually`: https://github.com/FormidableLabs/terraform-provider-serverless/issues/3

**Comment by tptee** [comment]
Added README!

**Pull request merged**